### DISCUSSION TITLE
How to use decorators inside constructor arguments

### DISCUSSION THREAD

**Discussion by Chiqui1234ok** [typegoose/typegoose#957] [Q&A] [answered]
Hi! I'm relatively new in Typescript, and I want to implement MongoDB with Typegoose (thanks for this piece of software! Seems pretty useful).
I saw an example in how to use _@prop_ with attributes defined inside a class. The example is this:

```typescript
class User {
  @prop()
  public name?: string;

  @prop({ required: true })
  public age!: number; // This is a single Primitive

  @prop({ type: () => [String] })
  public preferences?: string[]; // This is a Primitive Array

  @prop()
  public mainJob?: Job; // This is a single SubDocument
```

But for ease of mind, I have my attributes defined inside the constructor's arguments, like this:

```typescript
constructor(
        protected name: Name,
        protected phone: Phone,
        protected email: Email,
        protected password: Password,
        protected thumbnail: string,
        protected userNotes: Note[],
        protected admNotes: Note[],
        protected roles: Role[],
        protected jwt: string
    ) {
        // ...
    }
```
This saves time, because I don't have to initialize attributes if it was written like that.

It's possible to use @prop() before "protected"? Because I can't:

```typescript
constructor(
        @prop({ type: Name, required: true }) protected name: Name,
        @prop({ type: Phone, required: true }) protected phone: Phone,
        @prop({ type: Email, required: true }) protected email: Email,
        // ...
    )
```

Thanks in advance!

**Comment by hasezoey** [comment]
`constructor` is somewhat special as it allows definition in types as well as automatically removing some boilerplate as you have there, but at least with `experimentalDecorators` they still count as parameters and so only get *Parameter Decorators* applied (typegoose's `@prop` is a [*Property Decorator*](https://www.typescriptlang.org/docs/handbook/decorators.html#property-decorators)).
it would likely be possible for typegoose to add that functionality, but i currently do not plan to add parameter decorator functionality.

PS: in typegoose you dont actually construct the classes (unless you want to use them as POJO's)

some technical notes:

the Input:
```ts
class User {
  @prop()
  public username?: string;

  constructor(
    @prop() public test?: string,
    public t?: any,
    @prop() public s?: number
  ) {}
}
```
(when ignoring the errors typescript gives)

compiles to:
```js
let User = class User {
    constructor(test, t, s) {
        this.test = test;
        this.t = t;
        this.s = s;
    }
};
__decorate([
    (0, typegoose_1.prop)(),
    __metadata("design:type", String)
], User.prototype, "username", void 0);
User = __decorate([
    __param(0, (0, typegoose_1.prop)()),
    __param(2, (0, typegoose_1.prop)()),
    __metadata("design:paramtypes", [String, Object, Number])
], User);
```

**Comment by Chiqui1234ok** [comment, marked as answer]
Just if somebody wants to stick with this type of constructor (which helps reducing boilerplate), I will upgrade Mongoose to 8.6.2 and implement the schema with my class. If you're in the same boat, you maybe want to check [Advanced Schemas](https://mongoosejs.com/docs/advanced_schemas.html). This implements *loadClass()* function. The loadClass() function lets you pull in methods, statics, and virtuals from an ES6 class.

Cheers!
